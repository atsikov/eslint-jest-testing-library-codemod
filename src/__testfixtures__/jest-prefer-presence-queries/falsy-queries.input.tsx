// @ts-nocheck
it('test', () => {
  expect(getByTestId('id')).toBeNull()
  expect(getByText('id')).toBeFalsy()
  expect(getByRole('id')).not.toBeInTheDocument()
  expect(getByLabelText('id')).not.toBeDefined()
  expect(screen.getByTestId('id')).toBeNull()
  expect(screen.getByText('id')).toBeFalsy()
  expect(screen.getByRole('id')).not.toBeInTheDocument()
  expect(screen.getByLabelText('id')).not.toBeDefined()
  expect(a.b.getByTestId('id')).toBeNull()
  expect(a.b.getByText('id')).toBeFalsy()
  expect(a.b.getByRole('id')).not.toBeInTheDocument()
  expect(a.b.getByLabelText('id')).not.toBeDefined()
})
