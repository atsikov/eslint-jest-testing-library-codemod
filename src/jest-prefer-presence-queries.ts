import {
  Identifier,
  Transform,
  JSCodeshift,
  MemberExpression,
  CallExpression,
} from 'jscodeshift'
import prettier from 'prettier'

const JEST_TRUTHY_ASSERTIONS = [
  '.not.toBeNull',
  '.toBeTruthy',
  '.not.toBeFalsy',
  '.toBeInTheDocument',
  '.toBeDefined',
]

const JEST_FALSY_ASSERTIONS = [
  '.toBeNull',
  '.not.toBeTruthy',
  '.toBeFalsy',
  '.not.toBeInTheDocument',
  '.not.toBeDefined',
]

const RTL_QUERIES = [
  'LabelText',
  'PlaceholderText',
  'Text',
  'AltText',
  'Title',
  'DisplayValue',
  'Role',
  'TestId',
]

const RTL_TRUTHY_QUERY_PREFIX = 'getBy'
const RTL_FALSY_QUERY_PREFIX = 'queryBy'

const RTL_QUERY_METHODS = [
  ...RTL_QUERIES.map(query => `${RTL_TRUTHY_QUERY_PREFIX}${query}`),
  ...RTL_QUERIES.map(query => `${RTL_FALSY_QUERY_PREFIX}${query}`),
]

const EXPECT_FUNCTION = 'expect'

function getCalleeName(j: JSCodeshift, node: CallExpression): string | null {
  let nameNode = node.callee
  while (nameNode && nameNode.type != 'Identifier') {
    if (nameNode.type === 'MemberExpression') {
      nameNode = (nameNode as MemberExpression).property
    } else {
      return null
    }
  }

  if (
    nameNode
    && nameNode.type === 'Identifier'
  ) {
    return nameNode.name
  }

  return null
}

const transform: Transform = (fileInfo, api) => {
  const j = api.jscodeshift
  const root = j(fileInfo.source)

  root
    .find(
      j.MemberExpression,
      {
        object: {
          type: 'CallExpression',
          callee: {
            type: 'Identifier',
            name: EXPECT_FUNCTION,
          },
        },
      },
    )
    .filter(path => (path.value.object as CallExpression).arguments.length === 1)
    .filter(path => {
      const { arguments: args } = path.value.object as CallExpression
      const firstArgument = args[0]
      // Callee could be an Identifier 'expect(getByText(...))' or member expression
      // 'expect(screen.getByText(...))'
      const calleeName = j(firstArgument).isOfType(j.CallExpression)
        && getCalleeName(j, firstArgument as CallExpression)
      return !!calleeName && RTL_QUERY_METHODS.includes(calleeName)
    })
    .forEach(path => {
      let jestAssertionName = `.${(path.value.property as Identifier).name}`
      let parentPath = path.parentPath
      while (
        j(parentPath).isOfType(j.MemberExpression)
        && j(parentPath.value.property).isOfType(j.Identifier)
      ) {
        const name = ((parentPath.value as MemberExpression).property as Identifier).name
        jestAssertionName += `.${name}`
        parentPath = parentPath.parentPath
      }

      const firstArgument = (path.value.object as CallExpression).arguments[0]
      const calleeName = j(firstArgument).isOfType(j.CallExpression)
        && getCalleeName(j, firstArgument as CallExpression)

      const calleeIdentifierPath = j(
        (path.value.object as CallExpression).arguments,
      )
        .find(
          j.Identifier,
          {
            name: calleeName,
          },
        )


      if (
        calleeName
        && calleeName.startsWith(RTL_TRUTHY_QUERY_PREFIX)
        && JEST_FALSY_ASSERTIONS.includes(jestAssertionName)
      ) {
        calleeIdentifierPath
          .replaceWith(
            j.identifier(
              `${RTL_FALSY_QUERY_PREFIX}${calleeName.slice(RTL_TRUTHY_QUERY_PREFIX.length)}`,
            ),
          )
      }

      if (
        calleeName
        && calleeName.startsWith(RTL_FALSY_QUERY_PREFIX)
        && JEST_TRUTHY_ASSERTIONS.includes(jestAssertionName)
      ) {
        calleeIdentifierPath
          .replaceWith(
            j.identifier(
              `${RTL_TRUTHY_QUERY_PREFIX}${calleeName.slice(RTL_FALSY_QUERY_PREFIX.length)}`,
            ),
          )
      }
    })

  return prettier.format(root.toSource(), {
    parser: 'typescript',
    singleQuote: true,
    semi: false,
    trailingComma: 'all',
    printWidth: 90,
    arrowParens: 'avoid',
  })
}

export const parser = 'tsx'
export default transform
