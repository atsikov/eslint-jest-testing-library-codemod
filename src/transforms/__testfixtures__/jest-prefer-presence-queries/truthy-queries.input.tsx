// @ts-nocheck
it('test', () => {
  expect(queryByTestId('id')).not.toBeNull()
  expect(queryByText('id')).toBeTruthy()
  expect(queryByRole('id')).toBeInTheDocument()
  expect(queryByLabelText('id')).toBeDefined()
  expect(screen.queryByTestId('id')).not.toBeNull()
  expect(screen.queryByText('id')).toBeTruthy()
  expect(screen.queryByRole('id')).toBeInTheDocument()
  expect(screen.queryByLabelText('id')).toBeDefined()
  expect(a.b.queryByTestId('id')).not.toBeNull()
  expect(a.b.queryByText('id')).toBeTruthy()
  expect(a.b.queryByRole('id')).toBeInTheDocument()
  expect(a.b.queryByLabelText('id')).toBeDefined()
})
