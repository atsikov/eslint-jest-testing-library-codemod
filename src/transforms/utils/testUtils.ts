// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore no such function in typings
import { applyTransform } from 'jscodeshift/dist/testUtils'
import { existsSync, readFileSync } from 'fs'
import { resolve } from 'path'
import { Transform, Options, FileInfo } from 'jscodeshift'
import { format } from 'prettier'

export const snapshotTest = (
  module: { default: Transform, parser?: string },
  options: Options,
  inputFileName: string,
  testName?: string,
): void => {
  it(testName || 'transforms correctly', () => {
    const transformed = applyTransform(module, options, getInput(inputFileName))
    const formatted = format(transformed, {
      parser: 'typescript',
      singleQuote: true,
      semi: false,
      trailingComma: 'all',
      printWidth: 90,
      arrowParens: 'avoid',
    })
    expect(formatted).toMatchSnapshot()
  })
}

function getInput(moduleName: string): FileInfo {
  const MODULE_EXTS = ['ts', 'tsx']
  const paths = MODULE_EXTS.map(ext =>
    resolve(__dirname, `../__testfixtures__/${moduleName}.input.${ext}`),
  )
  const fixturePath = paths.find(path => existsSync(path))
  if (!fixturePath) {
    throw new Error(`Unable to find fixture ${moduleName}`)
  }

  return {
    path: fixturePath,
    source: readFileSync(fixturePath).toString(),
  }
}
