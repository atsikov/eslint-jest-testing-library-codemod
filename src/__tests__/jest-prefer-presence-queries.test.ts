import * as transform from '../jest-prefer-presence-queries'
import { snapshotTest } from '../testUtils'

snapshotTest(
  transform,
  {},
  'jest-prefer-presence-queries/truthy-queries',
  'Use getBy in truthy queries',
)
snapshotTest(
  transform,
  {},
  'jest-prefer-presence-queries/falsy-queries',
  'Use queryBy in falsy queries',
)
