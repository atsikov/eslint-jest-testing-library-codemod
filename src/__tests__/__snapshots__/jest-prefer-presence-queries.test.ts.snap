// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Use getBy in truthy queries 1`] = `
"// @ts-nocheck
it('test', () => {
  expect(getByTestId('id')).not.toBeNull()
  expect(getByText('id')).toBeTruthy()
  expect(getByRole('id')).toBeInTheDocument()
  expect(getByLabelText('id')).toBeDefined()
  expect(screen.getByTestId('id')).not.toBeNull()
  expect(screen.getByText('id')).toBeTruthy()
  expect(screen.getByRole('id')).toBeInTheDocument()
  expect(screen.getByLabelText('id')).toBeDefined()
  expect(a.b.getByTestId('id')).not.toBeNull()
  expect(a.b.getByText('id')).toBeTruthy()
  expect(a.b.getByRole('id')).toBeInTheDocument()
  expect(a.b.getByLabelText('id')).toBeDefined()
})
"
`;

exports[`Use queryBy in falsy queries 1`] = `
"// @ts-nocheck
it('test', () => {
  expect(queryByTestId('id')).toBeNull()
  expect(queryByText('id')).toBeFalsy()
  expect(queryByRole('id')).not.toBeInTheDocument()
  expect(queryByLabelText('id')).not.toBeDefined()
  expect(screen.queryByTestId('id')).toBeNull()
  expect(screen.queryByText('id')).toBeFalsy()
  expect(screen.queryByRole('id')).not.toBeInTheDocument()
  expect(screen.queryByLabelText('id')).not.toBeDefined()
  expect(a.b.queryByTestId('id')).toBeNull()
  expect(a.b.queryByText('id')).toBeFalsy()
  expect(a.b.queryByRole('id')).not.toBeInTheDocument()
  expect(a.b.queryByLabelText('id')).not.toBeDefined()
})
"
`;
